atc/scheduler/runner.go

```diff
@@ -88,27 +93,37 @@ func (s *schedulerRunner) schedulePipeline(ctx context.Context, errGroup *multie
 		jobsMap[job.Name()] = job.ID()
 	}
 
-	for _, job := range jobsToSchedule {
-		schedulingLock, acquired, err := job.AcquireSchedulingLock(logger)
-		if err != nil {
-			return fmt.Errorf("acquire job lock: %w", err)
-		}
-
-		if !acquired {
+	for _, j := range jobsToSchedule {
+		if _, exists := s.running.LoadOrStore(j.ID(), true); exists {
+			// already scheduling this job
 			continue
 		}
 
-		// shadow loop var for the goroutine closure
-		job := job
-
 		s.guardJobScheduling <- struct{}{}
-		errGroup.Go(func() error {
+
+		jLog := logger.Session("job", lager.Data{"job": j.Name()})
+
+		go func(job db.Job) {
 			defer func() {
 				<-s.guardJobScheduling
+				s.running.Delete(job.ID())
 			}()
 
-			return s.scheduleJob(ctx, logger, schedulingLock, pipeline, job, resources, jobsMap)
-		})
+			schedulingLock, acquired, err := job.AcquireSchedulingLock(logger)
+			if err != nil {
+				jLog.Error("failed-to-acquire-lock", err)
+				return
+			}
+
+			if !acquired {
+				return
+			}
+
+			err = s.scheduleJob(ctx, logger, schedulingLock, pipeline, job, resources, jobsMap)
```

Comment:
Can we defer release the schedulingLock so we don't need to pass it in here?

Reply:
sure can!